### DISCUSSION TITLE
How to "pivot" data

### DISCUSSION THREAD

**Discussion by txjmb** [grafana/grafana-infinity-datasource#387] [Q&A]
I have data in the following form:

```
[
    {
        "timestamp": "2022-09-01T00:00:05Z",
        "keyname": "foo",
        "value": "20.1"
    },
    {
        "timestamp": "2022-09-01T00:00:05Z",
        "keyname": "bar",
        "value": "20.5"
    },
    {
        "timestamp": "2022-09-01T00:00:10Z",
        "keyname": "foo",
        "value": "20.9"
    },
    {
        "timestamp": "2022-09-01T00:00:10Z",
        "keyname": "bar",
        "value": "10.3"
    }
]
```

I need separate series for each keyname for grafana to use it correctly:

| timestamp | foo | bar |
| --- | --- | --- |
| 2022-09-01T00:00:05Z | 20.1 | 20.5 |
| 2022-09-01T00:00:10Z | 20.9 | 10.3 |

How using your plugin can I do this pivot UQL or otherwise?

**Comment by yesoreyeram** [comment]
I don't remember having pivot in UQL. But what you need is [this](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22timeseries%22,%22url%22:%22https:%2F%2Fapi.moogsoft.ai%2Fv1%2Falerts%3Flimit%3D1000%26sortBy%3Dfirst_event_time%26sortOrder%3Ddesc%26start%3D0%26utcOffset%3DGMT-0%26filter%3Dfirst_event_time%20%3E%20$%7B__from%7D%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%5B%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22timestamp%5C%22:%20%5C%222022-09-01T00:00:05Z%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22keyname%5C%22:%20%5C%22foo%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22value%5C%22:%20%5C%2220.1%5C%22%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22timestamp%5C%22:%20%5C%222022-09-01T00:00:05Z%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22keyname%5C%22:%20%5C%22bar%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22value%5C%22:%20%5C%2220.5%5C%22%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22timestamp%5C%22:%20%5C%222022-09-01T00:00:10Z%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22keyname%5C%22:%20%5C%22foo%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22value%5C%22:%20%5C%2220.9%5C%22%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22timestamp%5C%22:%20%5C%222022-09-01T00:00:10Z%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22keyname%5C%22:%20%5C%22bar%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22value%5C%22:%20%5C%2210.3%5C%22%5Cn%20%20%20%20%7D%5Cn%5D%22,%22uql%22:%22parse-json%5Cn%7C%20extend%20%5C%22timestamp%5C%22%3Dtodatetime(%5C%22timestamp%5C%22)%22%7D%5D,%22range%22:%7B%22from%22:%221661990405000%22,%22to%22:%221661990410000%22%7D%7D) ( the following query and format the result as **timeseries** format )

```kusto
parse-json
| extend "timestamp"=todatetime("timestamp")
```

<img width="1791" alt="image" src="https://user-images.githubusercontent.com/153843/188598188-fc0bf70b-4881-451d-8926-b40dc25ccab0.png">

<img width="1731" alt="image" src="https://user-images.githubusercontent.com/153843/188598438-2d1601ab-3968-4026-8ce9-5c757b2b078e.png">

**Comment by txjmb** [comment]
Thank you for your response.  The issue in your example is that the values are not merged onto the same timestamp.  How would that be done in UQL?  In other words, there is a row per value/key for each timestamp, where there should only be two rows in a true pivot on the timestamp.  I'm not sure that Grafana will have any issues with this output from a graphing perspective, but it will not work well in a table format.

**Reply by yesoreyeram** [reply]
Have you tried summarize method? Refer docs for example

**Reply by yesoreyeram** [reply]
Grafana shouldn't have any issues withe the format i shared

**Reply by txjmb** [reply]
Yes, Grafana won't have any issue with it.  I am also trying to display in tabular form, for which I will have to use the summarize function, as you've said.  I'll give that a try.  Thank you!

**Reply by txjmb** [reply]
How do you escape quotes in UQL in column names?  In the results from the extend, the column name is `value {keyname="import__revenue", product="-"}`.  I want to do something like `summarize "Revenue"=sum("{keyname=\"import__revenue\", product=\"-\"}")` for that column name, but it has quotes in it and doubling the double-quotes doesn't seem to work, nor backslashes.
